docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,185 @@
+
+== Prerequisites
+
+In order to use the filter, a compatible KMS (Key Management Service) must be provided.  The KMS is used to store
+the encryption keys used to encrypt the records sent to the topic.  The service is used at runtime. If the KMS is
+unavailable, it will become impossible to produce or consume records through the filter until the KMS service is
+restored.  It is recommended to use a KMS in a highly available configuration.
+
+Currently, there is a single KMS integration: https://www.hashicorp.com/[HashiCorp Vault&#174;].  More KMS integrations
+are planned soon.
+
+It is also possible for a user to implement their own KMS integration.  The implementation must implement the
+{kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the implementation available on
+the classpath with the service loader mechanism.
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
+3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.
+
+These steps are described in the next sections.
+
+=== Setting up the KMS
+
+In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.
+
+==== HashiCorp Vault
+
+The filter integrates with the https://developer.hashicorp.com/vault/docs/secrets/transit[HashiCorp Vault *transit
+engine*].   Vault does not enable the transit engine by default.  It must be
+https://developer.hashicorp.com/vault/docs/secrets/transit#setup[enabled] before it can be used with the filter.
+
+The transit engine's path must be `/transit` (the default).
+
+The filter's configuration requires:
+
+* the HashiCorp Vault endpoint
+* a https://developer.hashicorp.com/vault/docs/concepts/tokens[token] that will be used by the filter to read the
+  encryption keys.
+
+=== Filter Configuration
+
+The filter is configured as part of the filter chain in the following way:
+
+[source, yaml]
+----
+filters:
+- type: EnvelopeEncryption                                        # <1>
+  config:
+    kms: KmsServiceName                                           # <2>
+    kmsConfig:                                                    # <3>
+      ..:
+    selector: SelectorName                                        # <4>
+    selectorConfig:                                               # <5>
+      ..:
+----
+<1> The name of the filter. This must be `EnvelopeEncryption`.
+<2> The KMS provider name.
+<3> Object providing configuration understood by KMS provider.
+<4> The key selector name.
+<5> Object providing configuration understood by key selector.
+
+==== KMS configuration
+
+For the KMS configuration:
+
+===== HashiCorp Vault
+
+For HashiCorp Vault, the KMS configuration looks like this:
+
+[source, yaml]
+----
+kms: VaultKmsService                                          # <1>
```

Comment:
This makes me wonder if we really need the `Service` in the name. Or `Kms` for that matter. Just like with Filters and FilterFactories the person doing the configuring doesn't really care about the class name. It makes no sense to them why there's `Service` in the name. To them the KMS is just `HashicorpVault`